### PR TITLE
Please help me write idiomatic Rust code 🙏

### PR DESCRIPTION
**Pull request by kossnocorp**
**TL;DR**: Please review my ISO 8601 parser. I'm looking for idiomatic code and performance comments, but any suggestions are welcome!

---

Hey!

I needed to write a flexible ISO 8601 parser, and both [chrono](https://crates.io/crates/chrono) and [iso8601](https://crates.io/crates/iso8601) didn't support partial strings like `2024`, `2024-01`, etc.

This is my naive implementation. I'm sure it misses many edge cases, but I'm generally pretty happy with the code as a foundation for further improvements.

I have limited Rust experience, so I do not yet have a sense of idiomatic Rust code. I would really appreciate any help on that front, but any comments are welcome. Additionally, I'm interested in making the code as performant as possible, and I think using the approach with mutable struct would be much faster than the chrono's way with using methods like [`with_year`](https://docs.rs/chrono/latest/chrono/trait.Datelike.html#tymethod.with_year) that return a new struct.

If you like, instead of sending a review, you can [reply to me on Twitter](#).

Thank you!

### CODE REVIEW

File: rust/iso-8601/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    month: u32,
+    date: u32,
+    hours: u32,
+    minutes: u32,
+    seconds: u32,
+    timezone: TimezoneValues,
+}
+
+impl DateTimeValues {
+    fn default() -> Self {
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:29]: In idiomatic rust this should be an implementation of the Default trait

File: rust/iso-8601/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            month: 1,
+            date: 1,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+            timezone: TimezoneValues::default(),
+        }
+    }
+
+    fn from(str: &str) -> Self {
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:41]: This should be From<AsRef<str>> so you can accept both &str and String

Comment (dmtrKovalenko) [reply rust/iso-8601/src/lib.rs:41]: If you want to support only str name it `from_str`

File: rust/iso-8601/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    Year(i32),
+    Month(u32),
+    Date(u32),
+    Hours(u32),
+    Minutes(u32),
+    Seconds(u32),
+    Timezone(TimezoneValues),
+}
+
+#[derive(Debug, PartialEq)]
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:93]: For empty enums you can always very cheaply define Eq as well

File: rust/iso-8601/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,296 @@
+use chrono::{DateTime, FixedOffset, TimeZone};
+
+pub fn parse_datetime(str: &str) -> Option<DateTime<FixedOffset>> {
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:3]: In rust it libraries especially it is not common to put functions in struct impls. 

So the uses of your libraries are concise and do not include a lot of random functions.

File: rust/iso-8601/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        .single()
+    })
+}
+
+struct DateTimeValues {
+    year: i32,
+    month: u32,
+    date: u32,
+    hours: u32,
+    minutes: u32,
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:23]: You can define minutes and seconds and hours as u8 and save a little bit bytes

File: rust/iso-8601/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    Month,
+    Date,
+    Hours,
+    Minutes,
+    Seconds,
+    Timezone,
+}
+
+impl ParserState {
+    fn next(&self) -> Option<Self> {
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:105]: This could be done as iterator and rust nerds will force you to do so but I think it's fine.

File: rust/iso-8601/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+            ParserState::Hours => str.parse().ok().map(|v| DateTimeField::Hours(v)),
+            ParserState::Minutes => str.parse().ok().map(|v| DateTimeField::Minutes(v)),
+            ParserState::Seconds => str.parse().ok().map(|v| DateTimeField::Seconds(v)),
+            _ => None,
+        }
+    }
+
+    fn parse_from(&self, str: &str) -> Option<usize> {
+        match self.prefix() {
+            Some(ParserPrefix::Dash) => str.get(..1).map(|s| if s == "-" { 1 } else { 0 }),
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:152]:
```suggestion
            Some(ParserPrefix::Dash) => if str.get(..1)? s == "-" { 1 } else { 0 },
```

File: rust/iso-8601/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+struct TimezoneValues {
+    sign: bool,
+    hours: i32,
+    minutes: i32,
+}
+
+impl TimezoneValues {
+    fn default() -> Self {
```
Comment (dmtrKovalenko) [inline rust/iso-8601/src/lib.rs:176]: Default trait

**Review by dmtrKovalenko — COMMENTED** [review]
Added a bunch of comments on the rust code idiomacity. Additionally: it is unclear which methods are public and which are private. 

Public structs should always derive Debug and methods should have correct conventions. 

A bit on alghoritm: this kind of parsers should target O(1) with no additional allocations. 

To make yourself more confident I can also recommend to setup clipping with pedantic preset. 

Hope my review was helpful.

**Comment by kossnocorp** [comment]
@dmtrKovalenko thank you a lot 🙏 It's very helpful. I don't understand everything, but I'll learn more about what you mentioned and apply it in the code. If you don't mind, I might have follow-up questions once I get to the code.
